tests/test_reactive.py

```diff
@@ -598,3 +598,70 @@ def compose(self) -> ComposeResult:
     app = MyApp()
     async with app.run_test():
         assert app.query_one(MyWidget).foo == "foobar"
+
+
+async def test_no_duplicate_external_watchers() -> None:
+    """Make sure we skip duplicated watchers."""
+
+    counter = 0
+
+    class Holder(Widget):
+        attr = var(None)
+
+    class MyApp(App[None]):
+        def __init__(self) -> None:
+            super().__init__()
+            self.holder = Holder()
+
+        def on_mount(self) -> None:
+            self.watch(self.holder, "attr", self.callback)
+            self.watch(self.holder, "attr", self.callback)
+
+        def callback(self) -> None:
+            nonlocal counter
+            counter += 1
+
+    app = MyApp()
+    async with app.run_test():
+        before = counter
```

Comment:
Any reason why you do the before and after stuff here instead of `assert counter == 1` before the assignment and then `assert counter == 2` after the assignment? That would seem simpler and a more accurate test.

As it stands, this test wouldn't catch duplicate watchers firing on `init`.

Reply:
I was testing whether the assignment triggered a single watcher but I agree your suggestion yields a more accurate test.

Reply:
See 1e682c2 for the change you suggested + a fix to a bug that your suggestion uncovered.

Reply:
Yup, looks good!